File: packages/protocol-schema/src/models/approval.ts
import { Type, type Static } from "@sinclair/typebox";

export const ApprovalKindSchema = Type.Union(
  [
    Type.Literal("command"),
    Type.Literal("file_change"),
    Type.Literal("user_input"),
  ],
  { $id: "ApprovalKind" }
);
export type ApprovalKind = Static<typeof ApprovalKindSchema>;

export const ApprovalStatusSchema = Type.Union(
  [
    Type.Literal("pending"),
    Type.Literal("approved"),
    Type.Literal("denied"),
    Type.Literal("canceled"),
  ],
  { $id: "ApprovalStatus" }
);
export type ApprovalStatus = Static<typeof ApprovalStatusSchema>;

export const ApprovalDecisionSchema = Type.Union(
  [
    Type.Literal("accept"),
    Type.Literal("accept_for_session"),
    Type.Literal("decline"),
    Type.Literal("cancel"),
  ],
  { $id: "ApprovalDecision" }
);
export type ApprovalDecision = Static<typeof ApprovalDecisionSchema>;

export const ApprovalPayloadSchema = Type.Object(
  {
    command: Type.Optional(Type.String()),
    cwd: Type.Optional(Type.String()),
    reason: Type.Optional(Type.String()),
    diff: Type.Optional(Type.String()),
    path: Type.Optional(Type.String()),
    prompt: Type.Optional(Type.String()),
    isHighRisk: Type.Optional(Type.Boolean()),
  },
  { $id: "ApprovalPayload" }
);
export type ApprovalPayload = Static<typeof ApprovalPayloadSchema>;

export const ApprovalRequestSchema = Type.Object(
  {
    id: Type.String({ format: "uuid" }),
    chatId: Type.String(),
    turnId: Type.String(),
    externalRequestId: Type.String(),
    kind: ApprovalKindSchema,
    payload: ApprovalPayloadSchema,
    status: ApprovalStatusSchema,
    requestedAt: Type.Integer(),
    resolvedAt: Type.Optional(Type.Integer()),
    resolvedByDeviceId: Type.Optional(Type.String()),
    resolvedByDeviceName: Type.Optional(Type.String()),
  },
  { $id: "ApprovalRequest" }
);
export type ApprovalRequest = Static<typeof ApprovalRequestSchema>;

export const ApprovalResponseInputSchema = Type.Object(
  {
    approvalId: Type.String(),
    decision: ApprovalDecisionSchema,
  },
  { $id: "ApprovalResponseInput" }
);
export type ApprovalResponseInput = Static<typeof ApprovalResponseInputSchema>;

File: crates/canywhere-protocol/src/models/approval.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalKind {
    Command,
    FileChange,
    UserInput,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalStatus {
    Pending,
    Approved,
    Denied,
    Canceled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalDecision {
    Accept,
    AcceptForSession,
    Decline,
    Cancel,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalPayload {
    pub command: Option<String>,
    pub cwd: Option<String>,
    pub reason: Option<String>,
    pub diff: Option<String>,
    pub path: Option<String>,
    pub prompt: Option<String>,
    pub is_high_risk: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalRequest {
    pub id: String,
    pub chat_id: String,
    pub turn_id: String,
    pub external_request_id: String,
    pub kind: ApprovalKind,
    pub payload: ApprovalPayload,
    pub status: ApprovalStatus,
    pub requested_at: i64,
    pub resolved_at: Option<i64>,
    pub resolved_by_device_id: Option<String>,
    pub resolved_by_device_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalResponseInput {
    pub approval_id: String,
    pub decision: ApprovalDecision,
}
